cmd/allocator/main.go

```diff
@@ -114,15 +120,15 @@ func main() {
 			select {
 			// watch for events
 			case event := <-watcherTLS.Events:
-				h.tlsMutex.Lock()
 				tlsCert, err := readTLSCert()
 				if err != nil {
 					logger.WithError(err).Error("could not load TLS cert; keeping old one")
 				} else {
+					h.tlsMutex.Lock()
+					defer h.tlsMutex.Unlock()
 					h.tlsCert = tlsCert
 				}
 				logger.Infof("Tls directory change event %v", event)
-				h.tlsMutex.Unlock()
 			case event := <-watcher.Events:
 				h.certMutex.Lock()
 				caCertPool, err := getCACertPool(certDir)
```

Comment:
[This is for line 203, but I can't comment there]

From https://github.com/googleforgames/agones/pull/1638

> Me: Since this line changes the value protected by the mutex, should the mutex be held here as well?
> You: No, newServiceHandler is called to create the serviceHandler, so no one can access the object yet.

Maybe it's not necessary in golang, but in other languages, setting a variable outside of a lock and then accessing it in a different thread (in a lock) doesn't guarantee that the second thread sees the value set in the first thread. This is because the lock is what instructs the program to introduce a memory fence to make sure that the value is visible across threads (e.g. move it outside of the cpu specific cache that isn't visible to other cpus on a multi-cpu machine).

Reply:
Yeah, I'm not exactly sure what Go's memory consistency model is (if it is sequentially consistent, I think this should be fine, but that's a pretty strong assumption). The ca cert initialization actually has the same problem, so I surrounded that with its corresponding lock as well. Fixed.